index.js

```diff
@@ -161,178 +100,238 @@ function invalidationsAsPatches(invalidated, currentTree, nextTree) {
   return patches;
 }
 
-Filter.prototype.build = function() {
-  // @ts-ignore
-  let srcDir = this.inputPaths[0];
-  // @ts-ignore
-  let destDir = this.outputPath;
+function invoke(context, fn, args) {
+  return new Promise(resolve => {
+    resolve(fn.apply(context, args));
+  });
+}
+
+class Filter extends Plugin {
+  constructor (inputTree, options) {
+    super([inputTree], {
+      name: (options && options.name),
+      annotation: (options && options.annotation),
+      persistentOutput: true
+    });
+    if (!this || !(this instanceof Filter) ||
```

Comment:
This cannot work here, it is invalid to call a class's constructor function directly. In fact this points to this change being a breaking change (because now `new` is required, but before the change it is not).

This scenario is supported today, but after changing the export to a class directly it will not be supported:

```js
const filter = require('broccoli-persistent-filter');

let tree = filter(someInputNode);
```

Reply:
Yeah. I realized it later. 
Below line was supposed to be removed if we are going ahead with `Class`
https://github.com/stefanpenner/broccoli-persistent-filter/blob/0311abb921878545a39b3fea7e2b62782a9f2fc0/index.js#L116

If we are going to support `let tree = filter(inputNode)`. We can not move to `Class`. Correct me if I am wrong. 

This change I made when I wanted to use the latest version of `broccoli-plugin@3.0.0` which moved to `Class` breaks the `broccoli-persistent-filter`. 
  Now I figured out that we just need to do a small change below 
https://github.com/stefanpenner/broccoli-persistent-filter/blob/7ba962659f2e6759b8e14e76253d9b83c0d32a96/index.js#L87
to
```
 Plugin.prototype.constructor.call(this, [inputTree], {
```
 fix those breaking changes.

Reply:
@rwjblue I think we still need this check to avoid usage 
```js
const Filter = require('broccoli-persistent-filter');
const testFilter = new Filter(inputNode);
```
Since broccoli-persistent-filter is a base class.
https://github.com/stefanpenner/broccoli-persistent-filter/blob/ef309604ade1c9ba0dedf528a96c4ab99132377e/test/test.js#L78

Reply:
Sorry, I think we must be talking past each other here somewhat. The thing I'm saying is that within a native `class`'s `constructor` function, it is **impossible** to have a `this` that isn't an instance of itself.

The two invocation styles that this library currently supports are:

```js
const Filter = require('broccoli-persistent-filter');

// ...snip...
return new Filter(someNode, { ...options }); 
```

And:

```js
const filter = require('broccoli-persistent-filter');

// ...snip...
return filter(someNode, { ...options });
```

If you want to continue to support both of those invocations after this change, you will need to do something _other than_ the check that these comments are associated with since doing the second example just above will throw an error.

Reply:
@stefanpenner Do we expect the second case mentioned by @rwjblue to be supported going forward?

Reply:
> The two invocation styles that this library currently supports are:
> ... snip ...
> In fact this points to this change being a breaking change (because now new is required, but before the change it is not).

@rwjblue I think your assertion that this library supports two invocation styles is incorrect. This code throws an error in the case where the class wasn't constructed with `new` and that was the behavior before @SparshithNR's patch. So on this issue, I don't think this patch breaks backwards compatibility. I'd recommend that we remove the `!this ||` conditional branch in the constructor check because it's now impossible for that code path to ever evaluate to true.

Reply:
@chriseppstein Yes, you are right. `!this` is an unnecessary check. I have removed it.

Reply:
@rwjblue Are we good with the changes?